lib/calculator_runner.rb

```diff
@@ -0,0 +1,75 @@
+class CalculatorRunner
+  def initialize
+    @calculators = []
+  end
+
+  def fetch_all!
+    while (calculator_class = next_calculator_with_dependencies_met)
+      puts "Running calculator: #{calculator_class}"
+      calculator_class
+        .new(
+          candidates: Candidate.all,
+          ballot_measures: Referendum.all,
+          committees: Committee.all
+        )
+        .fetch
+      @calculators.delete(calculator_class)
+    end
+
+    if @calculators.any?
+      error = "Could not meet dependencies for calculators:\n" +
+        @calculators.map do |c|
+          "  #{c} (Unmet dependencies: " + unmet_dependencies(c).map do |d|
+            "#{d[:model].name} #{d[:calculation]}"
+          end.join(", ") + ")"
+        end.join("\n")
+      raise error
+    end
+  end
+
+  def load_calculators(calculator_directory)
+    # load calculators dynamically, assume each one defines a class given by its
+    # filename. E.g. calculators/foo_calculator.rb would define "FooCalculator"
+    Dir.glob(calculator_directory).each do |calculator_file|
+      basename = File.basename(calculator_file.chomp('.rb'))
+      class_name = ActiveSupport::Inflector.classify(basename)
+      begin
+        calculator_class = class_name.constantize
+        @calculators << calculator_class
+      rescue NameError => ex
+        if ex.message =~ /uninitialized constant #{class_name}/
+          $stderr.puts "ERROR: Undefined constant #{class_name}, expected it to be "\
+            "defined in #{calculator_file}"
+          puts ex.message
+          exit 1
+        else
+          raise
+        end
+      end
+    end
+
+    self
+  end
+
+  private
+
+  def next_calculator_with_dependencies_met
+    @calculators.find do |calculator_class|
```

Comment:
Not that this probably takes much processing, but it would be more efficient if we sorted the list with all those with no dependencies first.  Even sorting by number of dependencies might make sense.

Reply:
Good idea, but I'm going to pass on it for now since this is going to be such a negligible amount of time saved, especially in comparison to the rest of the process.